### DISCUSSION TITLE
Understanding `BaseXCom.clear` when working with external data stores (Custom xcom backend)

### DISCUSSION THREAD

**Discussion by jmaldon1** [apache/airflow#20029] [Q&A]
From my current understanding this is how an Airflow DAG works.

(Clear Task A XCOM) Task A -> (Clear Task B XCOM) Task B -> (Clear Task C XCOM) Task C

I overrode the `BaseXCom` class to send XComs to a blob storage (S3, azure blob storage, etc). [Basic example of what I am doing](https://www.astronomer.io/guides/custom-xcom-backends#setting-up-an-xcom-backend)
I would like to delete this blob data at the end of the DAG (Lets say I have sensitive data that I don't want sitting in blob storage for longer than the life of the DAG). How do I go about deleting this data?

@dstandish @MatrixManAtYrService @sighingnow

**Comment by dstandish** [comment]
Various ways to skin this cat.

Probably something like this would be your simplest solution (though probably this code would not work as is):

Write an operator similar to this...

```python
class XcomCleanupOperator(BaseOperator):
    def execute(self, context: Any):
        ti = context['ti']  # type: TaskInstance
        xcom_values = ti.xcom_pull() # maybe needs more params like task_ids
        hook = S3Hook()
        for s3_uri in xcom_values:
            hook.delete_key(s3_uri)
```

Then add this as the last task in your dag.

So at the end it wipes all the files.

**Reply by jmaldon1** [reply]
Ah yes, was trying to avoid having to add an extra task at the end of each DAG (because we may have tons of DAGs storing things to blob storage, and I didn't want them all to have to append this extra task at the end). I was hoping there was a way to do it within the `BaseXCom` but I am guessing its not really possible.

@dstandish The only flaw with your approach is that if the DAG were to fail midway and never recover, it would have data sitting in blob storage forever.

Another approach is to create a DAG dedicated to deleting old files from blob storage, but then we are not tied to the DAG lifecycle anymore :/

**Comment by sighingnow** [comment]
I think you could a have hook in the DAG to clean the sensitive data.

See also https://stackoverflow.com/questions/44360853/airflow-run-task-regardless-of-upstream-success-fail

You could have `on_failure_callback/on_success_callback` in `default_args` as well to execute a callback for each task in a DAG, see also https://github.com/apache/airflow/blob/main/airflow/example_dags/tutorial.py
